### PR TITLE
Enable Mark as Watched in all the other playlist fragments.

### LINKED ISSUE #7159: Mark videos as watched in all places
### Checklist

- [x] I checked, but didn't find any duplicates (open OR closed) of this issue in the repo. 
- [X] I have read the contribution guidelines given at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md.
- [X] This issue contains only one feature request. I will open one issue for every feature I want to request.

#### Describe the feature you want

We recently implemented the ability to mark a video as watched, but this option is only visible from the subscriptions view. We want to make it work from any feed.

#### Is your feature request related to a problem? Please describe it

I would like to be able to mark videos as watched in a playlist, for example.

#### Additional context

This was actually originally suggested by @Pentaphon here: https://github.com/TeamNewPipe/NewPipe/pull/6773#issuecomment-906786123

I intend to implement this shortly.

#### How will you/everyone benefit from this feature?

We'll be able to mark videos as watched without being subscribed to their feed.

### PR DESCRIPTION
**Pull request by nschulzke**
#### What is it?
- [ ] Bugfix (user facing)
- [X] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

- Enable "Mark as Watched" in all the non-feed fragments

#### Before/After Screenshots/Screen Record

- Before:
Screenshot_1632455279

- After:
Screenshot_1632455130

#### Fixes the following issue(s)

- Fixes #7159

#### APK testing 

The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [X] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nschulzke** [comment]
Just a note: while testing this, I found that when I tapped "Mark as Watched" in several of the fragments, it did not immediately update. These same fragments also do not immediately update their visual status when you watch a video the regular way, so I didn't attempt to fix the reactivity in this pull request.

File: app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java [hunk 1]
```diff
@@ -366,6 +367,19 @@ private void showStreamDialog(final StreamStatisticsEntry item) {
...
         }
+
+        // show "mark as watched" only when watch history is enabled
+        final boolean isWatchHistoryEnabled = PreferenceManager
+                .getDefaultSharedPreferences(context)
+                .getBoolean(getString(R.string.enable_watch_history_key), false);
+        if (item.getStreamEntity().getStreamType() != StreamType.AUDIO_LIVE_STREAM
+                && item.getStreamEntity().getStreamType() != StreamType.LIVE_STREAM
+                && isWatchHistoryEnabled
+        ) {
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java:378]: This code should definitely be deduplicated. I'd create a static method `shouldAddMarkAsWatched(StreamType)` to `StreamDialogEntry` that replaces all of these lines and is also used in `FeedFragment`.

File: app/src/main/java/org/schabi/newpipe/local/history/HistoryRecordManager.java [hunk 2]
```diff
@@ -120,19 +120,11 @@ public HistoryRecordManager(final Context context) {
...
-                        streamId,
-                        duration * 1000
-                );
-                streamStateTable.insert(entity);
-            }
+            final StreamStateEntity entity = new StreamStateEntity(
+                    streamId,
+                    duration * 1000
+            );
+            streamStateTable.upsert(entity);
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/local/history/HistoryRecordManager.java:127]: Nice catch ;-)

**Review by Stypox — CHANGES_REQUESTED** [review]
> These same fragments also do not immediately update their visual status when you watch a video the regular way, so I didn't attempt to fix the reactivity in this pull request.

Yeah, that's fine

File: app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java (hunk 1 above)
Comment (nschulzke) [reply app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java]: Fair enough, that's done. I had to pass the context in as well to get `isWatchHistoryEnabled` working.

File: app/src/main/java/org/schabi/newpipe/local/history/HistoryRecordManager.java (hunk 2 above)
Comment (nschulzke) [reply app/src/main/java/org/schabi/newpipe/local/history/HistoryRecordManager.java:127]: Yeah, I didn't think to look for an `upsert` method the first time. XD

**Review by Stypox — APPROVED** [review]
Code looks good; I tested the APK in all of the fragments you changed and it seems to work fine. Thank you!

**Pull request merged**